doc/compiler/GPU/README.md

```diff
@@ -0,0 +1,228 @@
+# GPU Code Generation and Execution with OpenJ9
+
+## Summary
+In this document, we describe the GPU code generation feature in OpenJ9. Essentially, the just-in-time (JIT) compiler can compile and optimize parallel loops expressed as [Java 8's parallel stream API](https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html) for GPU execution as well as multi-core CPU execution. Currently, IBM POWER CPUs (ppc64le) and Intel CPUs (x86_64) plus [NVIDIA GPUs](https://developer.nvidia.com/cuda-gpus) are supported.
+
+## How it works
+The GPU code generation feature is implemeted in OpenJ9's JIT compiler. Consider the following Java program:
+```
+void heavy() {
+  IntStream.range(0, N).parallel().forEach(i -> {...});
+}
+```
+
+The ```heavy()``` method contains Java's Parallel Stream API and is at first interpreted on the JVM. The JVM periodically monitors the *hotness* of the method, and, when needed, generates an optimized native CPU program. The JIT compiler at first performs a limited set of compiler optimizations over intermediate representation (IR) so as not to hinder the main program execution. However, depending on the hotness, it may recompile the method with more aggressive optimizations including GPU code generation. The figure below illustrates an overview of the GPU code generation:
+
+![An overview of the GPU code generation.](./figs/gpucodegen.png "An overview of the GPU code generation." )
```

Comment:
We prefer not to check in binary files into the repository, otherwise things start to bloat very quick for everyone cloning the repository. For images such as this one we can use ASCII art however. I've taken your image and created an ASCII version which we can use instead of checking in the binary PNG:

```
   +----------------+
   | Java Bytecode  |
   +----------------+
           |
+----------|---------------------------------------------------------------------------------------+
|          v                                                                                       |
|  +----------------+       +----------------+       +----------------+       +----------------+   |   +-----------+
|  |  Translation   | +---> |    Existing    | +---> |Parallel Streams| +---> |Target Machine  | +---> |    CPU    |
|  |   to our IR    |       | Optimizations  |       | Identification |   +-> |Code Generation |   |   |   Binary  |
|  +----------------+       +----------------+       +----------------+   |   +----------------+   |   +-----------+
|                                                            |            |                        |   |GPU Runtime|
|                                                            v            |                        |   +-----------+
|                                                    +----------------+   |                        |
|                                                    |  Optimization  +---+                        |
|                                                    |    for GPUs    |                            |
| Our JIT Compiler                                   +----------------+                            |
|                                                            |                                     |
+--------------------------------------------------------------------------------------------------+
                                                             |
+------------------------------------------------------------v-------------------------------------+
|                                                    +----------------+       +----------------+   |   +-----------+
|                                                    |    NVVM IR     | +---> |    libNVVM     | +---> |    GPU    |
| NVIDIA's Tool Chain                                |                |       |                |   |   |   Binary  |
|                                                    +----------------+       +----------------+   |   +-----------+
+--------------------------------------------------------------------------------------------------+

```

Reply:
@fjeremic thank you very much! I'll definitely use the ASCII version.

Reply:
For reference I've created the above using ASCIIFlow [1] which is a very handy tool. Thanks for your contribution!

[1] http://asciiflow.com/